File: L1S/v3/code/ocaml/src/ir.ml
type ir_type = I32 | I64 | U32 | U64 | I8 | U8 | Bool | Void | Ptr of ir_type

type vreg = int

type ir_instr =
  | Const of vreg * int64 * ir_type
  | Add of vreg * vreg * vreg
  | Sub of vreg * vreg * vreg
  | Mul of vreg * vreg * vreg
  | Div of vreg * vreg * vreg
  | Eq of vreg * vreg * vreg
  | Lt of vreg * vreg * vreg
  | And of vreg * vreg * vreg
  | Or of vreg * vreg * vreg
  | Not of vreg * vreg
  | Neg of vreg * vreg
  | Load of vreg * vreg * ir_type
  | Store of vreg * vreg * ir_type
  | Call of vreg * string * vreg list * ir_type
  | Alloc of vreg * ir_type

type ir_terminator =
  | Ret of vreg
  | Br of string
  | BrCond of vreg * string * string

type ir_block = { label: string; instrs: ir_instr list; terminator: ir_terminator }

type ir_function = {
  name: string;
  params: ir_type list;
  return_type: ir_type;
  blocks: ir_block list;
  mutable num_vregs: int;
  slot_map: (int, int) Hashtbl.t;
  frame_size: int;
}

type ir_module = { funcs: ir_function list; globals: (string * ir_type) list }

let fresh_vreg func =
  let v = func.num_vregs in
  func.num_vregs <- func.num_vregs + 1;
  v

let make_block label = { label; instrs = []; terminator = Ret 0 }

let make_function name ret_type =
  { name; params = []; return_type = ret_type; blocks = []; num_vregs = 0;
    slot_map = Hashtbl.create 16; frame_size = 0 }

let ir_type_to_string = function
  | I32 -> "i32" | I64 -> "i64" | U32 -> "u32" | U64 -> "u64"
  | I8 -> "i8" | U8 -> "u8" | Bool -> "bool" | Void -> "void"
  | Ptr t -> Printf.sprintf "@%s" (ir_type_to_string t)

File: L1S/v3/code/rust/src/ir.rs
pub type VReg = usize;

#[derive(Clone, Copy, PartialEq, Debug)]
pub enum IrType {
    I32, I64, U32, U64, I8, U8, Bool, Void,
    Ptr(usize),
}

#[derive(Clone, Debug)]
pub struct IrModule {
    pub funcs: Vec<IrFunction>,
}

#[derive(Clone, Debug)]
pub struct IrFunction {
    pub name: String,
    pub param_types: Vec<IrType>,
    pub param_vregs: Vec<VReg>,
    pub return_type: IrType,
    pub blocks: Vec<IrBlock>,
    pub num_vregs: usize,
}

#[derive(Clone, Debug)]
pub struct IrBlock {
    pub label: String,
    pub instrs: Vec<IrInstr>,
    pub terminator: Option<IrTerminator>,
}

#[derive(Clone, Debug)]
pub enum IrInstr {
    Const(VReg, i64, IrType),
    Mov(VReg, VReg),
    Add(VReg, VReg, VReg),
    Sub(VReg, VReg, VReg),
    Mul(VReg, VReg, VReg),
    Div(VReg, VReg, VReg),
    Eq(VReg, VReg, VReg),
    Lt(VReg, VReg, VReg),
    And(VReg, VReg, VReg),
    Or(VReg, VReg, VReg),
    Not(VReg, VReg),
    Neg(VReg, VReg),
    Call(VReg, String, Vec<VReg>, IrType),
}

#[derive(Clone, Debug)]
pub enum IrTerminator {
    Ret(Option<VReg>),
    Br(String),
    BrCond(VReg, String, String),
}
